plugins/filter/lib/filter.cpp

```diff
@@ -0,0 +1,197 @@
+            return NO_FILTER;
+        } else if (m_rejected.exactMatch(number)) {
+            return FILTER_REJECT;
+        } else if (m_ignored.exactMatch(number)) {
+            return FILTER_IGNORE;
+        } else if (m_whitelist.match(number)) {
+            return NO_FILTER;
+        } else if (m_rejected.match(number)) {
+            return FILTER_REJECT;
+        } else if (m_ignored.match(number)) {
+            return FILTER_IGNORE;
+        } else {
+            return NO_FILTER;
+        }
+    }
+
+    FilterList m_rejected;
+    FilterList m_ignored;
+    FilterList m_whitelist;
+};
+
+Filter::Filter(QObject *parent)
+    : QObject(parent)
+    , d(new Private)
+{
+    connect(&d->m_rejected, &FilterList::changed,
+            this, &Filter::rejectedListChanged);
+    connect(&d->m_ignored, &FilterList::changed,
+            this, &Filter::ignoredListChanged);
+    connect(&d->m_whitelist, &FilterList::changed,
+            this, &Filter::whiteListChanged);
+}
+
+Filter::~Filter()
+{
+}
+
+QStringList Filter::rejectedList() const
+{
+    return d->m_rejected.list();
+}
+
+QStringList Filter::ignoredList() const
+{
+    return d->m_ignored.list();
+}
+
+QStringList Filter::whiteList() const
+{
+    return d->m_whitelist.list();
+}
+
+void Filter::ignoreNumber(const QString &number)
+{
+    d->m_whitelist.removeEntry(number);
+    d->m_rejected.removeEntry(number);
+    d->m_ignored.addEntry(number);
+}
+
+void Filter::ignoreNumbersStartingWith(const QString &prefix)
+{
+    QString pattern = prefix;
+    ignoreNumber(pattern.prepend('^'));
+}
+
+void Filter::ignoreByDefault()
+{
+    ignoreNumber(QStringLiteral("*"));
+}
+
+void Filter::rejectNumber(const QString &number)
+{
+    d->m_whitelist.removeEntry(number);
+    d->m_ignored.removeEntry(number);
+    d->m_rejected.addEntry(number);
+}
+
+void Filter::rejectNumbersStartingWith(const QString &prefix)
+{
+    QString pattern = prefix;
+    rejectNumber(pattern.prepend('^'));
+}
+
+void Filter::rejectByDefault()
+{
+    rejectNumber(QStringLiteral("*"));
+}
+
+void Filter::acceptNumber(const QString &number)
+{
+    d->m_rejected.removeEntry(number);
+    d->m_ignored.removeEntry(number);
+    if (d->m_rejected.match(number)
+        || d->m_ignored.match(number)) {
+        // Whitelist a number only if it is blocked by a pattern.
+        d->m_whitelist.addEntry(number);
+    }
+}
+
+void Filter::acceptNumbersStartingWith(const QString &prefix)
+{
+    QString pattern = prefix;
+    pattern.prepend('^');
+    d->m_ignored.removeEntry(pattern);
+    d->m_rejected.removeEntry(pattern);
+    if (d->m_rejected.match(prefix)
+        || d->m_ignored.match(prefix)) {
+        // Whitelist a prefix only if it is blocked by a larger pattern.
+        d->m_whitelist.addEntry(pattern);
+    }
+}
+
+void Filter::acceptByDefault()
+{
+    d->m_ignored.removeEntry(QStringLiteral("*"));
+    d->m_rejected.removeEntry(QStringLiteral("*"));
+}
+
+void Filter::acceptAll()
```

Comment:
clearAll() / clear() / clearFilters() or something to highlight it removes everything set?

Reply:
Well, I don't know if it's a good idea or not, but I tried to name the methods in a way that is independant from the actual three dconf list implementation. After a call to acceptAll(), I don't want to know what happens internally, but all calls are now accepted. If it's implemented with a "*" in the whitelist, leaving all previous filters or by removing the filter is an implementation detail and should not be reflected in the API. All methods should setup the internal lists to do what their names mean, which I think is the case at the moment.

But maybe it's not a good idea… Not sure if on the opposite, the methods should reflect exactly what internally they are doing and let the users create something valid from it. With my approach, I think it's easier to create a UI which can directly call the library methods : want to block a number ? call blockNumber(). want to block a range call blockNumbersStartingWith(), want to unblock a number ? call acceptNumber()…